Closes #20582. Non-executable transactions should not be evicted.

villanuevawill opened the pull request:
Closes #20582.

# Description
Queued transactions should not be evicted until after `evictionInterval` has passed. Successful promotion of transactions via `promoteExecutables`, pending replacement via `add(...)`, or insertion into the queue via `enqueueTx` should update the last heartbeat. The heartbeat is no longer removed when the `pending` list is emptied or after a minute has passed in the future queue.

Tests are expanded to ensure queued transactions are not removed before the `evictionInterval` duration has passed. Additionally, tests are expanded to ensure a reset or emptied pending list does not clear out the heart beat.

Logging is expanded to detect evictions as a method to prevent/monitor issues in the future.

# Monitoring Before & After
We see a **5x reduction** in queue evictions and the queue is able to reach capacity. Previously, I was seeing 5 tx/s evicted from the queue, that has been reduced to 1 tx/s evicted. Additionally, the queue is able to grow to its limit, from ~500 to 1024 in the case of my config.

**Before**:
![image](https://user-images.githubusercontent.com/7415822/82740082-58531200-9d13-11ea-8665-cf0d83478321.png)
![image](https://user-images.githubusercontent.com/7415822/82740084-6739c480-9d13-11ea-854e-94eda4b2f0f6.png)

**After**:
![image](https://user-images.githubusercontent.com/7415822/82740087-73258680-9d13-11ea-9598-cc29580033a5.png)
![image](https://user-images.githubusercontent.com/7415822/82740092-7caeee80-9d13-11ea-816e-ca0fa22bb782.png)

# Questions
There could be a case for removing heartbeat updates on pending replaced and in promote executables, however, it seems optimal to keep the queue fresh when new activity is seen and to manage ordering appropriately on truncation.

villanuevawill commented:
@karalabe issue #20582 was originally assigned to you it looks like, so tagging you on here. I've been deep in txpool code due to our work on account abstraction, so wanted to fix issues along the way. lmk if anything you'd like updated here.

holiman commented:
@karalabe - I'm thinking it may be a good idea to spin this up on one of the bootnodes. Want me to do that?

karalabe commented:
@holiman Pls do. The PR seems tiny enough, will try to review tomorrow.

holiman commented:
I'll rebase it first

holiman commented:
`aws-eu-central`, last six hours before redeploying with this PR
![Screenshot_2020-05-27 Single Geth - Grafana](https://user-images.githubusercontent.com/142290/83052096-21506980-a04f-11ea-9bfe-7c9619b57658.png)

holiman commented:
After switching to this PR, 15 minutes: 
![Screenshot_2020-05-27 Single Geth - Grafana(2)](https://user-images.githubusercontent.com/142290/83062465-fcfc8900-a05e-11ea-89ab-df4369acfe09.png)

Here are a couple of hours, capturing both before and after switch

![Screenshot_2020-05-27 Single Geth - Grafana(3)](https://user-images.githubusercontent.com/142290/83062487-084fb480-a05f-11ea-864c-8a77a634557b.png)

Zooming in a bit

![Screenshot_2020-05-27 Single Geth - Grafana(4)](https://user-images.githubusercontent.com/142290/83062548-23222900-a05f-11ea-9164-caac2b388555.png)

holiman commented:
Not sure why this graph looks like this though: (same time-segment as above)
![Screenshot_2020-05-27 Single Geth - Grafana(5)](https://user-images.githubusercontent.com/142290/83062776-75634a00-a05f-11ea-8ff3-f75b849f5f42.png)

holiman commented:
This PR lgtm, has been running stable for a while now

rjl493456442 commented:
the `beats` map is a little confusing for me. It seems to represent the `Last heartbeat from each known account`(from the comment) but actually it's not in the current code or PR. 

Now we use `beats` for two things: (1) pick the least active users to evict queue transaction (2) drop old enough queue transactions. For the latter one, I think the best solution may record the timestamp of each transaction. 

And for the former, it's even worse. We only record the beat when we successfully promote a transaction. But if a user caches lots of queue transaction but never successfully promote any one. Then the queue transactions belong to this user will never be dropped.

holiman commented:
Another comparison chart, this PR (green) against another bootnode: 
![Screenshot_2020-06-04 Multi Geth - Grafana](https://user-images.githubusercontent.com/142290/83737920-f15b2480-a653-11ea-9313-7b9ca5d0a71a.png)

villanuevawill commented:
> Now we use `beats` for two things: (1) pick the least active users to evict queue transaction (2) drop old enough queue transactions. For the latter one, I think the best solution may record the timestamp of each transaction(but tbh not sure whether it's worthwhile to add this complexity).

@rjl493456442 `beats` are also used when the queue reaches capacity to determine which transactions to throw out - https://github.com/ethereum/go-ethereum/blob/master/core/tx_pool.go#L1341

< For the latter one, I think the best solution may record the timestamp of each transaction(but tbh not sure whether it's worthwhile to add this complexity).

Yes, this would be ideal and also agree it isn't worth the complexity at this moment.

< We only record the beat when we successfully promote a transaction. But if a user caches lots of queue transaction but never successfully promote any one. Then the queue transactions belong to this user will never be dropped.

I'm confused by this statement. This PR updates the beat when a transaction is promoted, enqueued or replaced/swapped for an existing pending transaction.

villanuevawill commented:
> For the latter one, I think the best solution may record the timestamp of each transaction(but tbh not sure whether it's worthwhile to add this complexity).

I'll put up an alternate PR this week with the optimized solution - it doesn't appear to add much complexity.

villanuevawill commented:
@holiman & @rjl493456442 I've made the updates described above and this is ready for review.

![image](https://user-images.githubusercontent.com/7415822/85358080-17941780-b4d0-11ea-9626-905d0dc1420e.png)
![image](https://user-images.githubusercontent.com/7415822/85358104-22e74300-b4d0-11ea-92cd-c8e4071d9022.png)
![image](https://user-images.githubusercontent.com/7415822/85358131-309cc880-b4d0-11ea-9d70-673b274244b7.png)

The pool looks stable locally with the updates. I've updated to:
1. Only update the heartbeat on promotion or pending replaced (not on enqueue)
2. Clear the heartbeat when pending is removed
3. Add a ts for each transaction hash
4. Evict a transaction from the queue **only** if the heartbeat is past the lifetime and the transaction is past the lifetime

I've added thorough testing to check the edge cases. Also, I updated `validateTxPoolInternals` in the tests to verify internals are correct and that the length of ts entries is always equal to the # of transactions in the queue: https://github.com/ethereum/go-ethereum/pull/21119/files#diff-59daeee0a5084a262da2e62280567af0R111 - this adds additional security to the changes and tests this thoroughly.

holiman commented on core/tx_pool.go:
```diff
@@ -363,7 +366,10 @@ func (pool *TxPool) loop() {
 				if time.Since(pool.beats[addr]) > pool.config.Lifetime {
 					for _, tx := range pool.queue[addr].Flatten() {
-						pool.removeTx(tx.Hash(), true)
+						if time.Since(pool.queued_ts[tx.Hash()]) > pool.config.Lifetime {
```
If we get a mismatch between what's in `pool.queued_tx` and other pools, the `pool.queued_ts[tx.Hash()]` will return `0`, and flush out the tx. I suppose that is indeed the correct way to handle that error. 
For testing though, it might be interesting to make it `panic`, and run it for a few days with a `panic` in there, just to see that we don't have any inconsistencies.

villanuevawill replied:
I just updated this to panic, thanks!

holiman replied:
Sorry, I was a bit unclear. We can't merge it to `master` with a `panic` in there, it was more of a note-to-self (or whoever puts it on a prod server) to put a panic check there.

holiman replied:
I'll drop the last commit and rebase it on master, ok?

villanuevawill commented on core/tx_pool.go:
```diff
@@ -363,7 +367,13 @@ func (pool *TxPool) loop() {
 				if time.Since(pool.beats[addr]) > pool.config.Lifetime {
 					for _, tx := range pool.queue[addr].Flatten() {
-						pool.removeTx(tx.Hash(), true)
+						if pool.queued_ts[tx.Hash()].IsZero() {
```
@holiman here we include the check and panic.

holiman commented on core/tx_pool.go:
```diff
@@ -231,11 +232,12 @@ type TxPool struct {
+	pending   map[common.Address]*txList   // All currently processable transactions
+	queue     map[common.Address]*txList   // Queued but non-processable transactions
+	beats     map[common.Address]time.Time // Last heartbeat from each known account
+	queued_ts map[common.Hash]time.Time    // Timestamp for when queued transactions were added
```
Nitpick: it's frowned upon in golang to use underscore -- go uses camelCase instead. Please change this

villanuevawill commented:
@holiman updates applied 💯

holiman commented:
Great! Did you by any chance run it with panics enabled yourself?

villanuevawill commented:
> Great! Did you by any chance run it with panics enabled yourself?

Yes I did for a while and it didn't panic.

holiman commented:
Closed in favour of https://github.com/ethereum/go-ethereum/pull/21300
